File: analysis/workspace.py
# Ethan Doughty
# workspace.py
"""Workspace scanning for multi-file MATLAB projects."""

from __future__ import annotations
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Tuple

from analysis.context import FunctionSignature
from ir import FunctionDef


@dataclass(frozen=True)
class ExternalSignature:
    """Signature of a function defined in an external .m file."""
    name: str           # Internal function name (from function declaration)
    param_count: int    # Number of input parameters
    return_count: int   # Number of output variables (0 = procedure)
    source_path: str    # Absolute path to the .m file (for diagnostics)


# Regex to extract first function signature from MATLAB source
# Handles 3 forms:
#   function [a, b] = name(params)    # multi-return
#   function result = name(params)    # single-return
#   function name(params)             # procedure
_FUNC_SIG_RE = re.compile(
    r'^\s*function\s+'
    r'(?:'
        r'\[([^\]]*)\]\s*=\s*(\w+)'   # multi-return: [a, b] = name
        r'|(\w+)\s*=\s*(\w+)'          # single-return: result = name
        r'|(\w+)'                        # procedure: name
    r')\s*(?:\(([^)]*)\))?',
    re.MULTILINE
)


def extract_function_signature(source: str) -> Optional[Tuple[str, int, int]]:
    """Extract the first function signature from MATLAB source code.

    Args:
        source: MATLAB source code as string

    Returns:
        (function_name, param_count, return_count) or None if no function found
    """
    match = _FUNC_SIG_RE.search(source)
    if not match:
        return None

    groups = match.groups()
    # groups: (multi_rets, multi_name, single_ret, single_name, proc_name, params)
    multi_rets, multi_name, single_ret, single_name, proc_name, params = groups

    if multi_rets is not None:
        # Multi-return form: [a, b] = name(...)
        func_name = multi_name
        return_vars = [v.strip() for v in re.split(r'[,\s]+', multi_rets) if v.strip()]
        return_count = len(return_vars)
    elif single_ret is not None:
        # Single-return form: result = name(...)
        func_name = single_name
        return_count = 1
    else:
        # Procedure form: name(...)
        func_name = proc_name
        return_count = 0

    # Extract param count (params is None when no parentheses present)
    if params is None:
        param_count = 0
    else:
        param_list = [p.strip() for p in params.split(',') if p.strip()]
        param_count = len(param_list)

    return (func_name, param_count, return_count)


def scan_workspace(directory: Path, exclude: str = None) -> Dict[str, ExternalSignature]:
    """Scan a directory for .m files and extract function signatures.

    Args:
        directory: Directory to scan
        exclude: Filename to exclude (typically the file being analyzed)

    Returns:
        Dict mapping filename stems to ExternalSignature objects
    """
    result = {}

    if not directory.is_dir():
        return result

    for file_path in directory.glob('*.m'):
        # Skip the excluded file (the one being analyzed)
        if exclude and file_path.name == exclude:
            continue

        try:
            source = file_path.read_text(encoding='utf-8', errors='replace')
            sig_tuple = extract_function_signature(source)
            if sig_tuple is None:
                # Script file or no function found
                continue

            func_name, param_count, return_count = sig_tuple
            # Key by filename stem (MATLAB dispatch semantics)
            key = file_path.stem
            result[key] = ExternalSignature(
                name=func_name,
                param_count=param_count,
                return_count=return_count,
                source_path=str(file_path.resolve())
            )
        except (OSError, UnicodeDecodeError):
            # File read failed; skip silently (conservative)
            continue

    return result


# Cache of parsed external files (keyed by absolute source_path)
_parsed_cache: Dict[str, object] = {}


def clear_parse_cache() -> None:
    """Clear the parsed external file cache. Useful for LSP cache invalidation."""
    _parsed_cache.clear()


def load_external_function(sig: ExternalSignature) -> Optional[Tuple[FunctionSignature, Dict[str, FunctionSignature]]]:
    """Load and parse an external .m file, extracting its primary function and subfunctions.

    Args:
        sig: ExternalSignature from workspace scanning

    Returns:
        (primary_FunctionSignature, subfunctions_dict) or None on any error
    """
    from frontend.pipeline import parse_syntax, lower_to_ir

    source_path = sig.source_path
    if source_path in _parsed_cache:
        ir_prog = _parsed_cache[source_path]
    else:
        try:
            source = Path(source_path).read_text(encoding='utf-8', errors='replace')
            syntax_ast = parse_syntax(source)
            ir_prog = lower_to_ir(syntax_ast)
            _parsed_cache[source_path] = ir_prog
        except Exception:
            return None

    primary = None
    subfunctions: Dict[str, FunctionSignature] = {}
    for item in ir_prog.body:
        if isinstance(item, FunctionDef):
            func_sig = FunctionSignature(
                name=item.name, params=item.params,
                output_vars=item.output_vars, body=item.body
            )
            if primary is None:
                primary = func_sig
            else:
                subfunctions[item.name] = func_sig

    if primary is None:
        return None

    return (primary, subfunctions)

File: fsharp/Workspace.fs
module Workspace

open System
open System.IO
open System.Text.RegularExpressions
open Context

// ---------------------------------------------------------------------------
// Workspace scanning for multi-file MATLAB projects.
// Port of analysis/workspace.py
// ---------------------------------------------------------------------------

// Regex to extract first function signature from MATLAB source.
// Handles 3 forms:
//   function [a, b] = name(params)    # multi-return
//   function result = name(params)    # single-return
//   function name(params)             # procedure
let private funcSigPattern =
    @"^\s*function\s+(?:\[([^\]]*)\]\s*=\s*(\w+)|(\w+)\s*=\s*(\w+)|(\w+))\s*(?:\(([^)]*)\))?"

let private funcSigRegex =
    Regex(funcSigPattern, RegexOptions.Multiline)


/// extractFunctionSignature: extract the first function signature from MATLAB source.
/// Returns (funcName, paramCount, returnCount) or None if no function found.
let extractFunctionSignature (source: string) : (string * int * int) option =
    let m = funcSigRegex.Match(source)
    if not m.Success then None
    else
        let multiRets  = m.Groups.[1].Value  // group 1: multi-return vars
        let multiName  = m.Groups.[2].Value  // group 2: multi-return func name
        let singleRet  = m.Groups.[3].Value  // group 3: single-return var
        let singleName = m.Groups.[4].Value  // group 4: single-return func name
        let procName   = m.Groups.[5].Value  // group 5: procedure func name
        let params_    = m.Groups.[6].Value  // group 6: parameter list

        let funcName, returnCount =
            if m.Groups.[1].Success then
                // Multi-return form: [a, b] = name(...)
                let returnVars =
                    multiRets.Split([| ','; ' '; '\t' |], StringSplitOptions.RemoveEmptyEntries)
                    |> Array.filter (fun s -> s.Trim() <> "")
                (multiName, returnVars.Length)
            elif m.Groups.[3].Success then
                // Single-return form: result = name(...)
                (singleName, 1)
            else
                // Procedure form: name(...)
                (procName, 0)

        let paramCount =
            if not m.Groups.[6].Success || params_.Trim() = "" then 0
            else
                params_.Split(',')
                |> Array.filter (fun s -> s.Trim() <> "")
                |> Array.length

        Some (funcName, paramCount, returnCount)


/// scanWorkspace: scan a directory for .m files and extract function signatures.
let scanWorkspace (dirPath: string) (excludeFile: string) : Map<string, ExternalSignature> =
    let result = System.Collections.Generic.Dictionary<string, ExternalSignature>()

    if not (Directory.Exists(dirPath)) then Map.empty
    else
        let mFiles = Directory.GetFiles(dirPath, "*.m")
        for filePath in mFiles do
            let fileName = Path.GetFileName(filePath)
            if fileName <> excludeFile then
                try
                    let source = File.ReadAllText(filePath)
                    match extractFunctionSignature source with
                    | None -> ()  // Script file or no function found
                    | Some (funcName, paramCount, returnCount) ->
                        let key = Path.GetFileNameWithoutExtension(filePath)
                        result.[key] <- {
                            filename    = funcName
                            paramCount  = paramCount
                            returnCount = returnCount
                            sourcePath  = filePath
                            body        = None
                            parmNames   = []
                            outputNames = []
                        }
                with _ -> ()  // File read failed; skip silently

        result |> Seq.map (fun kv -> (kv.Key, kv.Value)) |> Map.ofSeq


// ---------------------------------------------------------------------------
// Parse cache: path -> (content_hash, primary_sig * subfunctions)
// ---------------------------------------------------------------------------

let private parsedCache =
    System.Collections.Generic.Dictionary<string, string * (FunctionSignature * Map<string, FunctionSignature>)>()


/// clearParseCache: clear the parsed external file cache.
let clearParseCache () =
    parsedCache.Clear()


/// contentHash: compute MD5 hash of a string.
let private contentHash (content: string) : string =
    use md5 = System.Security.Cryptography.MD5.Create()
    let bytes = System.Text.Encoding.UTF8.GetBytes(content)
    let hash = md5.ComputeHash(bytes)
    hash |> Array.map (fun b -> b.ToString("x2")) |> String.concat ""


/// loadExternalFunctionFromPath: load and parse an external .m file given its full path.
/// This is the real implementation; called from analysis when we have the path.
let loadExternalFunctionFromPath
    (sourcePath: string)
    (parseAndBuildIr: string -> (FunctionSignature * Map<string, FunctionSignature>) option)
    : (FunctionSignature * Map<string, FunctionSignature>) option =

    try
        let source = File.ReadAllText(sourcePath)
        let hash = contentHash source

        match parsedCache.TryGetValue(sourcePath) with
        | true, (cachedHash, result) when cachedHash = hash -> Some result
        | _ ->
            match parseAndBuildIr source with
            | None -> None
            | Some result ->
                parsedCache.[sourcePath] <- (hash, result)
                Some result
    with _ -> None


/// buildIrFromSource: parse MATLAB source and extract function signatures.
/// Returns (primary_FunctionSignature, subfunctions_dict) or None if no function found.
let private buildIrFromSource (source: string) : (FunctionSignature * Map<string, FunctionSignature>) option =
    try
        let program = Parser.parseMATLAB source
        let funcDefs =
            program.body
            |> List.choose (fun stmt ->
                match stmt with
                | Ir.FunctionDef(_, _, name, parms, outputVars, body) ->
                    Some { name = name; parms = parms; outputVars = outputVars; body = body }
                | _ -> None)
        match funcDefs with
        | [] -> None
        | primary :: rest ->
            let subfunctions =
                rest |> List.map (fun s -> (s.name, s)) |> Map.ofList
            Some (primary, subfunctions)
    with _ -> None


/// loadExternalFunction: load and parse an external .m file.
/// Returns (primary_FunctionSignature, subfunctions_dict) or None on error.
let loadExternalFunction (sig_: ExternalSignature) : (FunctionSignature * Map<string, FunctionSignature>) option =
    if sig_.sourcePath = "" then None
    else loadExternalFunctionFromPath sig_.sourcePath buildIrFromSource
